pyquil/quilbase.py

```diff
@@ -477,86 +447,29 @@ class DefGate(AbstractInstruction):
     :param parameters: list of parameters that are used in this gate
     """
 
-    def __init__(
-        self,
+    def __new__(
+        cls,
         name: str,
-        matrix: Union[List[List[Any]], np.ndarray, np.matrix],
-        parameters: Optional[List[Parameter]] = None,
-    ):
-        if not isinstance(name, str):
-            raise TypeError("Gate name must be a string")
-
-        if name in RESERVED_WORDS:
-            raise ValueError("Cannot use {} for a gate name since it's a reserved word".format(name))
-
-        if isinstance(matrix, list):
-            rows = len(matrix)
-            if not all([len(row) == rows for row in matrix]):
-                raise ValueError("Matrix must be square.")
-        elif isinstance(matrix, (np.ndarray, np.matrix)):
-            rows, cols = matrix.shape
-            if rows != cols:
-                raise ValueError("Matrix must be square.")
-        else:
-            raise TypeError("Matrix argument must be a list or NumPy array/matrix")
-
-        if 0 != rows & (rows - 1):
-            raise ValueError("Dimension of matrix must be a power of 2, got {0}".format(rows))
-        self.name = name
-        self.matrix = np.asarray(matrix)
-
-        if parameters:
-            if not isinstance(parameters, list):
-                raise TypeError("Paramaters must be a list")
+        matrix: Union[List[List[Expression]], np.ndarray, np.matrix],
+        parameters: Optional[List[Parameter]],
+    ) -> "DefGate":
+        specification = DefGate._convert_to_matrix_specification(matrix)
+        rs_parameters = [param.name for param in parameters or []]
+        return super().__new__(cls, name, rs_parameters, specification)
 
-            expressions = [elem for row in self.matrix for elem in row if isinstance(elem, Expression)]
-            used_params = {param for exp in expressions for param in _contained_parameters(exp)}
-
-            if set(parameters) != used_params:
-                raise ValueError(
-                    "Parameters list does not match parameters actually used in gate matrix:\n"
-                    "Parameters in argument: {}, Parameters in matrix: {}".format(parameters, used_params)
-                )
-        else:
-            is_unitary = np.allclose(np.eye(rows), self.matrix.dot(self.matrix.T.conj()))
-            if not is_unitary:
-                raise ValueError("Matrix must be unitary.")
```

Comment:
Will need to add this to `quil-rs`. We don't check dimensions or if a matrix is unitary when defining a matrix gate specification.

Reply:
[Issue](https://github.com/rigetti/quil-rs/issues/178)